File: ranking_algorithm.py
import json
from get_location import get_gps_from_address
from haversine import haversine, Unit

def ranking_algorithm(restaurants, preferences, location):
    restaurantScores = []
    
    for restaurant in restaurants:
        totalScore = 0
        restaurantPrice = len(restaurant['Price'])
        restaurantDistance = calculateDistance(location, restaurant['Address'])

        for user in preferences:
            if restaurantPrice >= user['Price'][0] and restaurantPrice <= user['Price'][1]:
                priceScore = 4
            else:
                priceScore = 0
            
            if restaurantDistance == -1:
                distanceScore = 0
            elif restaurantDistance <= (user['Distance']*0.000621371):
                distanceScore = 3
            else:
                distanceScore = 0
            
            if user['Cuisine'] in restaurant['Cuisine']:
                cuisineScore = 2
            else:
                cuisineScore = 0

            if float(restaurant['Rating']) >= user['Rating']:
                ratingScore = 1
            else:
                ratingScore = 0
            
            # print(restaurant['Name'])
            # print(user['Name'])
            # print(priceScore)
            # print(distanceScore)
            # print(cuisineScore)
            # print(ratingScore)
            totalScore += priceScore + distanceScore + cuisineScore + ratingScore
            # totalScore += (priceScore * PRICE_WEIGHT) + (distanceScore * DISTANCE_WEIGHT) + (cuisineScore * CUISINE_WEIGHT) + (ratingScore * RATING_WEIGHT)
        
        restaurantScores.append((restaurant, totalScore))
        # print(restaurantScores)
    
    # Sort the restaurants based on total score in descending order
    sortedRestaurants = sorted(restaurantScores, key=lambda x: x[1], reverse=True)
    [print(restaurant[0]['Name'], restaurant[1]) for restaurant in sortedRestaurants]
    
    return sortedRestaurants

def calculateDistance(location, endAddress):
    if '#' in endAddress:
        temp = endAddress.split()
        endAddress = ' '.join([string for string in temp if '#' not in string])
    
    lat1, lon1 = float(location.split(',')[0]), float(location.split(',')[1])  # get_gps_from_address(location)
    try:
        lat2, lon2 = get_gps_from_address(endAddress)
    except:
        return 0

    # Calculate distance
    distance = haversine((lat1, lon1), (lat2, lon2), unit=Unit.MILES)

    return distance


def main():
    with open('exampleOutputGoogleAPI.json', 'r') as file:
        restaurants = json.load(file)
        # print(restaurants)
        # print("Type:", type(restaurants))
    
    with open('exampleInputGoogleAPI.json', 'r') as file:
        preferences = json.load(file)
        # print(preferences)
        # print("Type:", type(preferences))
    location  = '42.7284,-73.6918'
    ranking_algorithm(restaurants, preferences, location)
    return

if __name__ == '__main__':
    main()

# price -> +4
# distance -> +3
# cuisine -> +2
# rating -> +1

# for restaurant in restaurants:
#   if restaurant price <= price of preference

File: ranking_algorithm.js
require('dotenv').config(); // For environment variables
const axios = require('axios');
const haversine = require('haversine');

// Load the API key from the .env file
const GOOGLE_MAPS_API_KEY = process.env.GOOGLE_MAPS_API_KEY;

// Function to get GPS coordinates from address
async function getGpsFromAddress(address) {
  const url = `https://maps.googleapis.com/maps/api/geocode/json?address=${encodeURIComponent(address)}&key=${GOOGLE_MAPS_API_KEY}`;

  try {
    const response = await axios.get(url);

    if (response.data.status === 'OK' && response.data.results.length > 0) {
      const location = response.data.results[0].geometry.location;
      return { lat: location.lat, lng: location.lng };
    } else {
      throw new Error(`Geocoding failed: ${response.data.status}`);
    }
  } catch (error) {
    console.error('Error in geocoding:', error.message);
    return null;
  }
}

// Ranking algorithm
async function rankingAlgorithm(restaurants, preferences, location) {
  const restaurantScores = [];

  for (const restaurant of restaurants) {
    let totalScore = 0;
    const restaurantPrice = restaurant.Price.length;
    const restaurantDistance = await calculateDistance(location, restaurant.Address);

    for (const user of preferences) {
      let priceScore = 0, distanceScore = 0, cuisineScore = 0, ratingScore = 0;

      if (restaurantPrice >= user.Price[0] && restaurantPrice <= user.Price[1]) {
        priceScore = 4;
      }

      if (restaurantDistance === -1) {
        distanceScore = 0;
      } else if (restaurantDistance <= (user.Distance * 0.000621371)) { // convert meters to miles
        distanceScore = 3;
      }

      if (restaurant.Cuisine.includes(user.Cuisine)) {
        cuisineScore = 2;
      }

      if (parseFloat(restaurant.Rating) >= user.Rating) {
        ratingScore = 1;
      }

      totalScore += priceScore + distanceScore + cuisineScore + ratingScore;
    }

    restaurantScores.push({ restaurant, totalScore });
  }

  // Sort restaurants based on total score in descending order
  const sortedRestaurants = restaurantScores.sort((a, b) => b.totalScore - a.totalScore);

  sortedRestaurants.forEach(({ restaurant, totalScore }) => {
    console.log(restaurant.Name, totalScore);
  });

  return sortedRestaurants;
}

// Function to calculate distance between two locations
async function calculateDistance(location, endAddress) {
  // Clean the address if it contains '#'
  if (endAddress.includes('#')) {
    const temp = endAddress.split(' ').filter(str => !str.includes('#'));
    endAddress = temp.join(' ');
  }

  const [lat1, lon1] = location.split(',').map(Number);

  try {
    const location2 = await getGpsFromAddress(endAddress);

    if (!location2) {
      return -1;
    }

    // Calculate distance using the haversine formula
    const start = { latitude: lat1, longitude: lon1 };
    const end = { latitude: location2.lat, longitude: location2.lng };

    const distance = haversine(start, end, { unit: 'mile' });
    return distance;
  } catch (error) {
    console.error('Error in calculating distance:', error.message);
    return -1;
  }
}

// Main function to load data and execute ranking algorithm
async function main() {
  const fs = require('fs');

  const restaurants = JSON.parse(fs.readFileSync('exampleOutputGoogleAPI.json', 'utf8'));
  const preferences = JSON.parse(fs.readFileSync('exampleInputGoogleAPI.json', 'utf8'));
  const location = '42.7284,-73.6918';

  const rankedRestaurants = await rankingAlgorithm(restaurants, preferences, location);
  
  // Save the ranked results
  fs.writeFileSync('ranked_restaurants.json', JSON.stringify(rankedRestaurants, null, 2));
}

// Run the main function
main();
